stl/inc/functional

```diff
@@ -1207,6 +1207,604 @@ _NODISCARD bool operator!=(nullptr_t, const function<_Fty>& _Other) noexcept {
+        is_nothrow_invocable_r_v<_Rx, _Vt const, _Types...>&& is_nothrow_invocable_r_v<_Rx, _Vt const&, _Types...>;
+
+    _Rx operator()(_Types... _Args) const noexcept {
+        return this->_Get_invoke()(this->_Data, _STD forward<_Types>(_Args)...);
+    }
+};
+
+template <class _Rx, class... _Types>
+class _Move_only_function_call<_Rx(_Types...) const& noexcept> : public _Move_only_function_base<_Rx, _Types...> {
+public:
+    using result_type = _Rx;
+
+    template <class _Vt>
+    using _VtInvQuals = _Vt const&;
+
+    template <class _Vt>
+    static constexpr bool _Is_callable_from = is_nothrow_invocable_r_v<_Rx, _Vt const&, _Types...>;
+
+    _Rx operator()(_Types... _Args) const& noexcept {
+        return this->_Get_invoke()(this->_Data, _STD forward<_Types>(_Args)...);
+    }
+};
+
+template <class _Rx, class... _Types>
+class _Move_only_function_call<_Rx(_Types...) const&& noexcept> : public _Move_only_function_base<_Rx, _Types...> {
+public:
+    using result_type = _Rx;
+
+    template <class _Vt>
+    using _VtInvQuals = _Vt const&&;
+
+    template <class _Vt>
+    static constexpr bool _Is_callable_from = is_nothrow_invocable_r_v<_Rx, _Vt const&&, _Types...>;
+
+    _Rx operator()(_Types... _Args) const&& noexcept {
+        return this->_Get_invoke()(this->_Data, _STD forward<_Types>(_Args)...);
+    }
+};
+#endif // __cpp_noexcept_function_type
+
+template <class _Fx>
+class move_only_function : private _Move_only_function_call<_Fx> {
+private:
+    using _Call = _Move_only_function_call<_Fx>;
+
+    // clang-format off
+    template <class _Fn>
+    static constexpr bool _Enable_one_arg_constructor =
+        !is_same_v<_Remove_cvref_t<_Fn>, move_only_function> &&
+        !_Is_specialization_v<_Remove_cvref_t<_Fn>, in_place_type_t> &&
+        _Call::template _Is_callable_from<decay_t<_Fn>>;
+
+    template <class _Fn, class... _CTypes>
+    static constexpr bool _Enable_in_place_constructor =
+        is_constructible_v<decay_t<_Fn>, _CTypes...> &&
+        _Call::template _Is_callable_from<decay_t<_Fn>>;
+
+    template <class _Fn, class _Ux, class... _CTypes>
+    static constexpr bool _Enable_in_place_list_constructor =
+        is_constructible_v<decay_t<_Fn>, initializer_list<_Ux>, _CTypes...> &&
+        _Call::template _Is_callable_from<decay_t<_Fn>>;
+    // clang-format on
+public:
+    using typename _Call::result_type;
+
+    move_only_function() noexcept {
+        this->_Construct_with_null();
+    }
+
+    move_only_function(nullptr_t) noexcept {
+        this->_Construct_with_null();
+    }
+
+    move_only_function(move_only_function&&) noexcept = default;
+
+    template <class _Fn, enable_if_t<_Enable_one_arg_constructor<_Fn>, int> = 0>
+    move_only_function(_Fn&& _Callable) {
+        using _Vt = decay_t<_Fn>;
+        static_assert(is_constructible_v<_Vt, _Fn>);
+
+        if constexpr (is_member_function_pointer_v<_Vt> || is_member_object_pointer_v<_Vt> || is_pointer_v<_Vt>) {

```

Comment:
This doesn't appear to handle N4901 [func.wrap.move.ctor]/8, /8.3: "*Postconditions:* `*this` has no target object if any of the following hold: [...] `remove_cvref_t<F>` is a specialization of the `move_only_function` class template, and `f` has no target object.", which additionally indicates missing test coverage.

This can happen, despite the /5.1 constraint that "`remove_cvref_t<F>` is not the same type as `move_only_function`", when the callable object `F` is a `move_only_function<DifferentSignature>`. For example, constructing `move_only_function<int (int)>` from `move_only_function<short (long long)>`. We have test coverage for `std::function` carefully exercising this case (which you don't have to copy exactly, but can look at for precedent).